core/src/main/java/org/apache/druid/data/input/FirehoseFactory.java

```diff
@@ -77,4 +77,14 @@ default boolean isSplittable()
   {
     return false;
   }
+
+  /**
+   * Some FirehoseFactory implementations require the indexing service's TaskToolbox to function.
+   * Use this method to provide it to them. The argument should always be a TaskToolbox; the
+   * method signature uses Object so that FirehoseFactories don't all have to be inside the
+   * indexing-service module.
+   */
+  default void setTaskToolbox(Object taskToolbox)
```

Comment:
I'm really not sure how we can make this better without huge refactoring.. Let's open an issue about it after this PR.